flutter_tools: Copy vendored frameworks from plugin podspecs in ios/macos-framework builds

MohammedTarigg opened the pull request:
Fixes #125530

When running `flutter build ios-framework` or `flutter build macos-framework`, vendored frameworks declared in plugin podspecs (via `s.vendored_frameworks`) were not being copied to the output directory.

This PR adds support for parsing plugin podspecs to find vendored_frameworks entries and copying them to the output directory. Single .framework files are wrapped into xcframeworks to match the output format.

Changes:
- Added `copyVendoredFrameworks` method to `BuildFrameworkCommand` base class
- Added `parseVendoredFrameworks` function to parse podspec Ruby files
- Called from both iOS and macOS framework build commands
- Added unit tests for the podspec parser

I'm happy to adjust the approach if there's a better way to handle this - particularly around the regex-based podspec parsing. Let me know if this looks reasonable or if you'd prefer a different strategy.

MohammedTarigg committed:
feat: Implement copying vendored frameworks from plugin podspecs into Darwin framework builds.

vashworth commented on packages/flutter_tools/lib/src/commands/build_ios_framework.dart:
```diff
@@ -180,6 +181,140 @@ abstract class BuildFrameworkCommand extends BuildSubCommand {
+    if (singleValue.isNotEmpty) {
+      results.add(singleValue);
+    }
+  } else if (arrayContent != null) {
```
```suggestion
  if (singleValue != null && singleValue.isNotEmpty) {
    results.add(singleValue);
  } else if (arrayContent != null) {
```

vashworth commented on packages/flutter_tools/lib/src/commands/build_ios_framework.dart:
```diff
@@ -180,6 +181,140 @@ abstract class BuildFrameworkCommand extends BuildSubCommand {
+    }
+  } else if (arrayContent != null) {
+    // Extract individual paths from the array content.
+    final pathPattern = RegExp(r'''["']([^"']+)["']''');
```
Podspec syntax allows wildcard patterns such as:

```
s.vendored_frameworks = '**/*.xcframework'
```

https://guides.cocoapods.org/syntax/podspec.html#group_file_patterns

MohammedTarigg committed:
Update packages/flutter_tools/lib/src/commands/build_ios_framework.dart

Co-authored-by: Victoria Ashworth <15619084+vashworth@users.noreply.github.com>

MohammedTarigg committed:
Refactor vendored frameworks to parse project.pbxproj instead of podspec

This change addresses reviewer feedback by:
- Replacing podspec regex parsing with project.pbxproj parsing using PlistParser
- This approach is more reliable because CocoaPods has already resolved all paths,
  wildcards, and platform-specific entries
- Handles edge cases like spec.ios.vendored_frameworks and wildcard patterns
  automatically
- Added comprehensive unit tests for parseVendoredFrameworksFromPbxproj

Note: This only copies frameworks from CocoaPods-based plugins. Swift Package
Manager support will be added in a separate command as mentioned by @vashworth.

vashworth commented:
Overall, I think this looks good, but we still need unit tests for `copyVendoredFrameworks`

MohammedTarigg committed:
Add unit tests for copyVendoredFrameworks and format files

vashworth reviewed (commented):
I checked this out locally and tested it with the `fair` plugin and found a few bugs. See comments

MohammedTarigg committed:
Fix vendored framework filter, path resolution, and copy behavior

- Use /Flutter.framework instead of Flutter.framework in the skip
  filter to avoid false matches on names like FairDynamicFlutter
- Strip ../../../ prefix from framework paths that CocoaPods adds
  for virtual groups (Development Pods)
- Copy .framework directly instead of wrapping in xcframework,
  matching how CocoaPods handles vendored frameworks
- Update and add tests for all three changes

MohammedTarigg committed:
Fix formatting in build_darwin_framework_test.dart

vashworth reviewed (approved):
LGTM! @MohammedTarigg Thanks for working through this with me!

okorohelijah reviewed (approved):
there is a merge conflict but LGTM

hellohuanlin commented on packages/flutter_tools/lib/src/commands/build_ios_framework.dart:
```diff
@@ -276,6 +277,179 @@ abstract class BuildFrameworkCommand extends BuildSubCommand {
+        if (!destination.existsSync()) {
+          globals.logger.printTrace('Copying vendored framework: $frameworkName');
+          copyDirectory(frameworkEntity, destination);
+        }
```
these 2 seems identical?

MohammedTarigg committed:
flutter_tools: dedupe vendored framework copy
